packages/utils/src/Defer.ts

```diff
@@ -0,0 +1,68 @@
+import { once } from 'lodash'
+
+type ResolveFn<T> = (value: T | PromiseLike<T>) => void
+
+type RejectFn = (reason?: Error) => void
+
+const noopExecutor = () => {}
+
+/**
+ * Deferred promise allowing external control of resolve / reject.
+ */
+export class Defer<T> extends Promise<T> {
+    private readonly resolveFn: ResolveFn<T>
+    private readonly rejectFn: RejectFn
+    private readonly ensureNoopCatchAttached: () => void
+    private settled = false
+
+    constructor(executor: (resolve: ResolveFn<T>, reject: RejectFn) => void = noopExecutor) {
```

Comment:
This cannot be left as an empty constructor otherwise invoking `resolve` or `reject` will result in `TypeError: Promise resolve or reject function is not callable`. See more discussion here https://github.com/nodejs/node/issues/13678